vault-plugins/flant_iam_auth/usecase/authn/accesstoken/validator.go

```diff
@@ -0,0 +1,319 @@
+				leeway = DefaultLeewaySeconds
+			}
+			*claims.NotBefore = jwt.NumericDate(int64(*claims.Expiry) - int64(leeway))
+		}
+	}
+
+	// Set clock skew leeway to apply when validating all time-related claims
+	cksLeeway := expected.ClockSkewLeeway
+	if expected.ClockSkewLeeway.Seconds() < 0 {
+		cksLeeway = 0
+	} else if expected.ClockSkewLeeway.Seconds() == 0 {
+		cksLeeway = jwt.DefaultLeeway
+	}
+
+	// Validate claims by asserting they're as expected
+	if (expected.Issuer != "" && expected.Issuer != claims.Issuer) || v.issuer != claims.Issuer {
+		return nil, fmt.Errorf("invalid issuer (iss) claim")
+	}
+	if expected.Subject != "" && expected.Subject != claims.Subject {
+		return nil, fmt.Errorf("invalid subject (sub) claim")
+	}
+	if expected.ID != "" && expected.ID != claims.ID {
+		return nil, fmt.Errorf("invalid ID (jti) claim")
+	}
+	if err := validateAudience(expected.Audiences, claims.Audience); err != nil {
+		return nil, fmt.Errorf("invalid audience (aud) claim: %w", err)
+	}
+
+	// Validate that the token is not expired with respect to the current time
+	now := time.Now()
+	if expected.Now != nil {
+		now = expected.Now()
+	}
+	if claims.NotBefore != nil && now.Add(cksLeeway).Before(claims.NotBefore.Time()) {
+		return nil, errors.New("invalid not before (nbf) claim: token not yet valid")
+	}
+	if claims.Expiry != nil && now.Add(-cksLeeway).After(claims.Expiry.Time()) {
+		return nil, errors.New("invalid expiration time (exp) claim: token is expired")
+	}
+	if claims.IssuedAt != nil && now.Add(cksLeeway).Before(claims.IssuedAt.Time()) {
+		return nil, errors.New("invalid issued at (iat) claim: token issued in the future")
+	}
+
+	return allClaims, nil
+}
+
+// enrichedClaims validates access_token, and returns enriched claims
+//
+// The given JWT is considered valid if:
+//  0. iss match issuer at OIDCDiscoveryURL
+//  1. oidc-provider returns userinfo presenting access_token
+//
+//  enrichedClaims returns claims given by the token and new claims from userinfoURL
+func (v *accessTokenValidator) enrichedClaims(ctx context.Context, token string) (map[string]interface{}, error) {
+	userinfo, err := v.getUserInfo(ctx, token)
+
+	parsedJWT, err := jwt.ParseSigned(token)
+	if err != nil {
+		return nil, err
+	}
+	allClaims := map[string]interface{}{}
+	err = parsedJWT.UnsafeClaimsWithoutVerification(&allClaims)
+	if err != nil {
+		return nil, err
+	}
+	for k, v := range userinfo {
+		if _, allredyExists := allClaims[k]; !allredyExists {
+			allClaims[k] = v
+		}
+	}
+	return allClaims, nil
+}
+
+func (v *accessTokenValidator) getUserInfo(ctx context.Context, token string) (map[string]interface{}, error) {
+	req, err := http.NewRequest(http.MethodGet, v.userinfoURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", "Bearer "+token)
+	resp, err := v.client.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
```

Comment:
same 

https://github.com/flant/negentropy/pull/267/files#r765493406

Reply:
read body, if not OK - return in error body to infuriate